### PR TITLE
Deprecation fixes and improvements

### LINKED ISSUE #151: Show ban durations
Currently if an account request ban is set to expire it doesn't show the expiration date. It should be added.

### LINKED ISSUE #160: Show last activity in alt detector warning and provide links for detected account requests
For example, it would say something like this:

> Warning
> The following accounts have made account requests from the same IP address as this user:
> * jvvg (January 1, 2022)

This would be helpful for knowing how relevant it is. Also for account requests it would be helpful to link to the most recent request.

### LINKED ISSUE #168: Stop using $wgUser
Looks like it's used in:
* ScratchConfirmAccountHooks.php
* email.php
* RequestPage.php (a whole lot)

### LINKED ISSUE #169: [Typo] in IP block message
There is a typo in ine translation source file

https://github.com/InternationalScratchWiki/scratch-confirmaccount-v3/blob/7d5dc5c0a42d19f5a9c1e3154bb3d76e75e43c02/i18n/en.json#L81

should be
```diff
- Your IP address has blocked from creating accounts. The reason given is: $1
+ Your IP address has been blocked from creating accounts. The reason given is: $1
```

### PR DESCRIPTION
**Pull request by apple502j**
Resolves #151
Resolves #160
Resolves #168
Resolves #169

## Breaking changes
- Requires MW1.35+.
- Several hooks have been renamed to avoid conflict. This affects other extensions with CA3 integration.

## Bugfixes
- Fixed a bug where block expiry option "existing" and "othertime" were incorrectly displayed in non-English locale.
- Fixed PHP warning being displayed while checking if a user is a New Scratcher.
- Fixed deprecation warning on PHP8.2 due to passing `null` to `strpos` in `SpecialConfirmAccounts`.
- Fixed a bug where password confirmation match used weak comparison. This does not affect actual login.
- Fixed a bug where session invalid error showed raw HTML.

## New features
- For account blocks with expiration, show the duration; this continues to be not displayed for infinite blocks.
- Alt account request warning now includes the link to the latest request, its request date and status.
- Alt account edit/Checkuser warning now includes the link to the user/talk/contrib and the last edit date.
- Alt account and checkuser warnings are now shown in descending order.

## Internationalization related changes
- New strings added: `scratch-confirmaccount-verification-project`, `scratch-confirmaccount-ip-warning-request-time`, `scratch-confirmaccount-ip-warning-checkuser-time`, `scratch-confirmaccount-user-blocked-duration`.
- `scratch-confirmaccount-vercode-explanation` no longer links to the project directly; the link is now constructed by code, to allow `target=_blank`. All locales have been updated.
- Removed unused string `scratch-confirmaccount-waiting-request`.
- Updated Japanese translation.
- Added new and missing `qqq` translation.

## Technical and internal changes
- Manifest schema version is now 2. This includes config object change. The file is also formatted.
- Added editorconfig and gitignore.
- Added typehint to some methods.
- Replaced `==` with `===`, etc.
- Removed unused variables.
- Replaced `wfGetDB` with `ILoadBalancer::getConnection`.
- Conditionally replaced `DB_MASTER` with `DB_PRIMARY`.
- Hook system completely rewritten to use `HookHandler`.
- Dependency injection is now used in several places.
- `accountRequestFormData` now returns HTML.
- Fixed session request ID sometimes being integer and sometimes string; this is now always string.
- Fixed a bug where `AccountRequestSubmitted` hook could fire multiple times.
- Fixed a bug with PHPDoc showing a wrong type.

### CODE REVIEW

File: src/SpecialRequestAccount.php [hunk 1]
```diff
@@ -409,13 +434,11 @@ function handleAuthenticationFormSubmission(IDatabase $dbr) {
...
 		}
 
-		$linkRenderer = $this->getLinkRenderer();
-
 		$username = $request->getText('username');
 		$password = $request->getText('password');
 
 		//see if there are any requests with the given password
-		$passwordFactory = MediaWikiServices::getInstance()->getPasswordFactory();
+		$passwordFactory = $this->passwordFactory;
```
Comment (jacob-g) [inline src/SpecialRequestAccount.php:441]: Notpicking a bit here but I think this line is unnecessary, you can just use `$this->passwordFactory` further on down

File: src/SpecialRequestAccount.php [hunk 2]
```diff
@@ -516,28 +539,26 @@ function handleSendConfirmEmailSubmission() {
...
-		$session = $request->getSession();
+
 		$this->setHeaders();
 		
 		$this->checkReadOnly();
 
 		if ($request->wasPosted()) {
 			return $this->handleFormSubmission();
-		} else if ($par == '') {
+		} else if ($par === null || $par === '') {
```
Comment (jacob-g) [inline src/SpecialRequestAccount.php:554]: Could we use `empty($par)` instead? Seems a bit cleaner

File: src/database/DatabaseInteractions.php [hunk 3]
```diff
@@ -384,18 +387,40 @@ function setRequestEmailConfirmed($request_id, IDatabase $dbw) {
...
+		[
+			'ORDER BY' => 'request_timestamp DESC'
+		]		
 	);
+
+	$entries = [];
+	$usernames = [];
+
+    foreach ($results as $row) {
+		if (in_array($row->request_username, $usernames)) continue;
```
Comment (jacob-g) [inline src/database/DatabaseInteractions.php:418]: Fix indentation

**Review by jacob-g — COMMENTED** [review]
Overall looks pretty good other than a few minor mostly cosmetic things I pointed out. I'll run this on my local (MW 1.39 and PHP 8.1) when I get the chance to make sure this all works.

Thank you for all the work you put into this, it really means a lot and makes my life so much easier.

**Review by jacob-g — COMMENTED** [review]
MediaWiki still tells me that the PersonalUrls hook is deprecated. I guess add a ticket to remove it once we do the real upgrade?

**Review by jacob-g — COMMENTED** [review]
image
Looks like the error display is no longer in the red box, we may need to migrate that styling.

Edit: never mind, this is an issue with SWS, will file a ticket there.

**Review by jacob-g — COMMENTED** [review]
Login intercept also isn't showing that nicely:
image

**Review by jacob-g — COMMENTED** [review]
Got this deprecation when queueing a job:
```
 Deprecated: Use of JobQueueGroup::singleton was deprecated in MediaWiki 1.37. [Called from handleRequestActionSubmission in /var/www/html/w/extensions/scratch-confirmaccount-v3/src/subpages/RequestPage.php at line 647] in /var/www/html/w/includes/debug/MWDebug.php on line 381
```

**Review by jacob-g — CHANGES_REQUESTED** [review]
Special:ConfirmAccounts fails for me with this when I try to view a request:
```
Error: Call to undefined method User::getCanonicalName()

Backtrace:

from /var/www/html/w/extensions/scratch-confirmaccount-v3/src/subpages/RequestPage.php(438)
#0 /var/www/html/w/extensions/scratch-confirmaccount-v3/src/subpages/RequestPage.php(516): requestCheckUserDisplay()
#1 /var/www/html/w/extensions/scratch-confirmaccount-v3/src/SpecialConfirmAccounts.php(451): requestPage()
#2 /var/www/html/w/includes/specialpage/SpecialPage.php(701): SpecialConfirmAccounts->execute()
#3 /var/www/html/w/includes/specialpage/SpecialPageFactory.php(1428): SpecialPage->run()
#4 /var/www/html/w/includes/MediaWiki.php(316): MediaWiki\SpecialPage\SpecialPageFactory->executePath()
#5 /var/www/html/w/includes/MediaWiki.php(904): MediaWiki->performRequest()
#6 /var/www/html/w/includes/MediaWiki.php(562): MediaWiki->main()
#7 /var/www/html/w/index.php(50): MediaWiki->run()
#8 /var/www/html/w/index.php(46): wfIndexMain()
#9 {main}
```

This is the end of my review for now. I will continue reviewing when this is fixed.

**Review by Kenny2github — CHANGES_REQUESTED** [review]

**Comment by apple502j** [comment]
`JobQueueGroup` will remain untouched (will fix after dropping 1.35) - no clean way to make it work for both. The replacement is added in 1.37.

**Comment by jacob-g** [comment]
@apple502j I put a new ticket (#171) in for removing `JobQueueGroup::singleton` for when we upgrade to 1.39 so we can still keep track of it.

**Comment by jacob-g** [comment]
Did a smoke test of the request flow, seems to work without any issues other than the remaining deprecations that will need to wait until we actually do the 1.39 upgrade.

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,4 @@
+.vscode
+.DS_Store
```
Comment (jacob-g) [inline .gitignore:2]:
```suggestion

```

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,4 @@
...
+Thumbs.db
```
Comment (jacob-g) [inline .gitignore:3]:
```suggestion

```

**Review by jacob-g — CHANGES_REQUESTED** [review]
Make the changes to the .gitignore and we should be good to go here. Thank you for all the work you put into this.

**Review by jacob-g — COMMENTED** [review]
At this point I think we're ready to go. Tomorrow I'll test this on staging (still on MW 1.35) to make absolutely sure it works with all our other stuff and won't break production when we deploy it. Stay tuned.

**Review by jacob-g — APPROVED** [review]
Tested on staging, looks good. Ready to merge.

**Review by Kenny2github — APPROVED** [review]
Did a rejection test on staging, LGTM

**Pull request merged**